src/harness/sas_test_harness.py

```diff
@@ -133,54 +133,70 @@ def getSasTestHarnessVersion(self):
     The run() method is an overridden method from thread class and it gets invoked
     when the thread is started using thread.start().
     """
-    logging.info('Started Test Harness Server:%s at %s', self.name, self.http_server_url)
-    self.server.serve_forever()
+
+    # Any exceptions in the server should be caught and handled without
+    # intervention and not cause it to stop. This is an added safeguard so
+    # any unhandled errors will result in:
+    #   1) An error being logged (Most important to diagnose problem).
+    #   2) An attempt to restart the server (allowing the test to proceed).
+    self.stopped = False
+    while not self.stopped:
+      logging.info('Started Test Harness Server: %s at %s', self.name,
+                   self.http_server_url)
+      try:
+        self.server.serve_forever()
+      except:
+        traceback.print_exc()
+        logging.error('HTTPServer %s at %s died on %s, restarting...',
+                      self.name, self.http_server_url,
+                      datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ'))
+    logging.info('Shutting down Test Harness Server:%s at %s', self.name,
+                 self.http_server_url)
 
   def shutdown(self):
     """This method is used to stop HTTPServer Socket."""
+    self.stopped = True
     self.server.shutdown()
     logging.info('Stopped Test Harness Server:%s', self.name)
 
 
   def cleanDumpFiles(self):
-    """Clean existing dumpfile if any."""
-    temp_dir_path = self.getDumpFilePath()
+    """Clean existing dump file if any."""
+    temp_dir_path = self.getDumpFileDirectory()
     map(os.remove, (os.path.join(temp_dir_path, file_path)
                     for file_path in os.listdir(temp_dir_path)
                     if os.path.exists(os.path.join(temp_dir_path, file_path))))
     os.rmdir(temp_dir_path)
     logging.info('all dump records generated by this testcase are cleaned successfully')
 
-  def __createFadObject(self, all_activity_dump):
+  def __createFadObject(self, all_activity_dump, fad_generation_time):
     """Creates the FAD object format and embed the active dump file object of different recordtype.
     """
     full_activity_dump = {}
     full_activity_dump['files'] = all_activity_dump
-    end_time = time.strftime(datetime.now().replace(microsecond=0).isoformat()) + 'Z'
-    full_activity_dump['generationDateTime'] = end_time
+    full_activity_dump['generationDateTime'] = fad_generation_time
     full_activity_dump['description'] = 'Load {} FAD to SAS UUT'.format(self.host_name)
     return full_activity_dump
 
-  def __createFadRecord(self, encoded_url, record):
+  def __createFadRecord(self, encoded_url, record, file_name):
     """ Creates the files object that includes url,checksum,size,version and recordtype of record
         required to be packed in FAD.
     """
     fad_record = {}
     fad_record['url'] = encoded_url
     fad_record['checksum'] = hashlib.sha1(json.dumps(record)).hexdigest()
```

Comment:
do we have the same issue with the checksum (need to calculate the checksum for the entire file, not just the record)?